### PR TITLE
Restore retries

### PR DESCRIPTION
**Pull request by dlau**
This commit reinstates the promise-retry dependency which provides
timeouts, exponential backoff, and retry functionality. In this
implementation, retry functionality is opt-in, meaning no retries will
be tried unless you explicitly request them.

We've wrapped promise-retry to handle all the heavy lifting of retry.
You can pass options to promise-retry (such as the exponent for
exponential backoff, and delay between attempts) by using a special
option `retry`.

Rather than constrain when to retry requests, we'll simply call the
function you provide inside the retry options `shouldRety` with the
FetchResponse or Error (such as ECONNRESET). If that
function returns true, we'll retry. If false, we resolve or reject.

Usage example (see readme for more):

```
jsonFetch('http://www.test.com/products/1234', {
  method: 'POST',
  body: {name: 'apple'},
  retry: {
    // Retry 5 times, in addition to the original request
    retries: 5,

    // after every request, decide whether or not to make another request
    shouldRetry: retries.isNetworkError
  }
}).then(response => {
  // handle 200-level responses:
});
```
cc @brycefisher  @goodeggs/shopping-web

### CODE REVIEW

**Comment by brycefisher** [comment]
cc @dannynelson I think this implementation matches what we discussed. Please chime with comments!

File: src/retriers.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    if(response && response.status &&
+      response.status >= 500 &&
+      response.status < 600) {
+        return true;
+    }
+    return false;
+  },
+
+  isNetworkError: (response) => {
+    return response instanceof Error;
```
Comment (dannynelson) [inline src/retriers.js]: The way `json-fetch` is currently implemented, this would retry for all non-200 responses. Is there a specific attribute we could look for on the error to be more specific?

File: test/test.js [hunk 2]
```diff
@@ -1,61 +1,210 @@
...
-  it('resolves with undefined for 404 status code', () => {
-    nock('http://www.test.com')
-      .get('/products/1234')
-      .reply(404, 'not found')
-    return jsonFetch('http://www.test.com/products/1234').then((response) => {
-      expect(response.body).to.equal(undefined)
-      expect(response.status).to.equal(404)
-      expect(response.statusText).to.equal('Not Found')
-      expect(response.headers).to.be.ok
+    it('resolves with undefined for 404 status code', () => {
```
Comment (dannynelson) [inline test/test.js]: Do you think 404s should error too? I think it makes sense error if an endpoint doesn't exist. We could return `null` if the endpoint does exist but it has an empty response.

File: src/retriers.js (hunk 1 above)
Comment (dlau) [inline src/retriers.js]: Fetch doesn't throw on non success codes
```
> fetch('https://github.com/goodeggs/fakenorepository404').then(() => {console.log('1')}).catch((err) => {console.log('2')});
> Promise {[[PromiseStatus]]: "pending", [[PromiseValue]]: undefined}
> GET https://github.com/goodeggs/fakenorepository404 404 (Not Found)(anonymous function) @ VM456:1
> 1
```

Comment (dannynelson) [inline src/retriers.js]: Oops! I thought we were calling it on the error thrown by json-fetch. It is ok then 👍

File: src/index.js [hunk 3]
```diff
@@ -1,21 +1,44 @@
...
+  return promiseRetry((retry) => {
+    return fetch(URL, fetchOptions)
+      .then((response) => {
+        if(retryOptions.shouldRetry(response)) {
+          return retry();
+        }
+        return response;
+      })
+      .catch((err) => {
+        if(retryOptions.shouldRetry(err)) {
```
Comment (brycefisher) [inline src/index.js]: Yeah! I had that thought too. If that feels better to you, I'm totally down with it. I think we did it this way initially since `shouldRetry` is part of this module's retry functionality...but your comment totally crossed my mind. I'm totally happy to pull it out into a top a level key.

File: src/retriers.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+export let retriers = {
+  is5xx: (response) => {
```
Comment (brycefisher) [inline src/retriers.js]: Probably! I think the intention here was to provide a few default retriers with some limited utility and also as documentation. Having said that, probably 503 / 504 would probably be the errors I'd actually want to retry as a consumer of this module.

File: test/test.js (hunk 2 above)
Comment (brycefisher) [inline test/test.js]: Hmm, we _could_. Some http libraries work that way -- throw an error for statusCodes >= 400. I think there's some real merit and precedence behind that approach.

My personal experience is that for some endpoints, a 200 status can contain an error or a 419 can be rate limiting, or other surprising status codes can have strange semantics in a given context. Because of those experiences, I tend to think of endpoint error semantics as highly application specific. In my personally preferred approach, any http response is a success for the http client and its the caller's responsibility (aka an SDK) to determine whether the http response represents a success or failure with its greater knowledge of the context.

So, @dlau and I simply reformated this line, and we didn't change the semantics of this test or the behavior of jsonFetch. If we did treat specific response status codes as failures, I expect that would be a major version bump -- which is totally fine!

I guess, the most intuitive answer to me is: "what does fetch do?" According to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Checking_that_the_fetch_was_successful):

> A fetch() promise will reject with a TypeError when a network error is encountered, although this usually means permission issues or similar — a 404 does not constitute a network error

I did a very quick look at node-fetch, and I found [this test case](https://github.com/bitinn/node-fetch/blob/master/test/test.js#L844-L849):

```js
fetch(url, opts).then(function(res) {
			expect(res.status).to.equal(404);
})
```

Anyway, I'm happy to chat in person in more depth if you'd like? Let me know!

File: src/index.js [hunk 5]
```diff
@@ -1,21 +1,44 @@
...
-  let fetchOptionsClone = objectAssign({}, fetchOptions);
+export {retriers} from './retriers';
+
+const DEFAULT_RETRY_OPTIONS = {
+  retries: 0,
+  shouldRetry: () => false
+};
+
+export default function jsonFetch(URL, options = {}) {
+  let fetchOptions = objectAssign({}, options);
```
Comment (brycefisher) [inline src/index.js]: You bet! I need to look at the best practices again -- do we prefer consts when fields on an object might change?

File: src/index.js [hunk 6]
```diff
@@ -1,21 +1,44 @@
...
   try {
-    fetchOptionsClone.body = JSON.stringify(fetchOptions.body);
+    fetchOptions.body = JSON.stringify(fetchOptions.body);
   } catch (err) {}
 
-  return fetch(URL, fetchOptionsClone)
+  return promiseRetry((retry) => {
+    return fetch(URL, fetchOptions)
+      .then((response) => {
+        if(retryOptions.shouldRetry(response)) {
```
Comment (brycefisher) [inline src/index.js]: I really wanted to add eslint here, but it felt like we'd already sunk a lot of time into getting this right.

File: test/test.js (hunk 2 above)
Comment (brycefisher) [inline test/test.js]: @dannynelson and I plan to chat about this in person later today and make a decision. I _think_ that's the only blocker to accepting this PR.

**Comment by brycefisher** [comment]
Many thanks to @dannynelson for the thorough and thoughtful review!

@goodeggs/ops-tools @goodeggs/shopping-web -- Any other input before we merge this today? If so, speak up soon! We hope to bump the version this morning if possible.

File: test/test.js (hunk 2 above)
Comment (dannynelson) [inline test/test.js]: @brycefisher and I decided to make this change as a separate pull request since it would be a breaking change.

**Comment by dannynelson** [comment]
Looks good! Thanks for making these changes 👍

File: test/test.js (hunk 2 above)
Comment (brycefisher) [inline test/test.js]: Recap of our inperson chat: @dannynelson proposal makes a lot of sense, but is a breaking change. He plans to make that change a separate PR to solicit feedback from the community. I'm going to merge this PR as is :-)

**Comment by demands** [comment]
Chiming in late, but this PR is awesome! Thanks @brycefisher @dlau 😺

**Comment by brycefisher** [comment]
Hooray! Thanks @demands
